src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py

```diff
@@ -369,29 +392,53 @@ def _purgeTree(self, ctx, baseDirPfn):
         :param ctx:  The gfal2 context object
         :return:     Bool: True if it managed to purge everything, False otherwise
         """
+        # NOTE: It deletes only directories and does not try to unlink any file.
+
         successList = []
 
+        # First test if baseDirPfn is actually a directory entry:
+        try:
+            entryStat = ctx.stat(baseDirPfn)
+        except gfal2.GError as gfalExc:
+            e = sys.exc_info()[1]
+            if e.code == errno.ENOENT:
+                self.logger.error("MISSING baseDir: %s", baseDirPfn)
+                successList.append(False)
+                return all(successList)
+            else:
+                self.logger.error("FAILED baseDir: %s: gfalException: %s", baseDirPfn, str(gfalExc))
+                successList.append(False)
+                return all(successList)
+        if not stat.S_ISDIR(entryStat.st_mode):
+            self.logger.error("The base pfn: %s is not a directory entry.", baseDirPfn)
+            successList.append(False)
+            return all(successList)
+
         if baseDirPfn[-1] != '/':
             baseDirPfn += '/'
 
+        # Second recursively iterate down the tree:
         for dirEntry in ctx.listdir(baseDirPfn):
             if dirEntry in ['.', '..']:
                 continue
             dirEntryPfn = baseDirPfn + dirEntry
             try:
                 entryStat = ctx.stat(dirEntryPfn)
-            except gfal2.GError:
+            except gfal2.GError as gfalExc:
                 e = sys.exc_info()[1]
                 if e.code == errno.ENOENT:
                     self.logger.error("MISSING dirEntry: %s", dirEntryPfn)
```

Comment:
@amaltaro I'd like to draw your attention to this exact error. 

What it says in the general case is: if we try to stat a missing directory while browsing the structure, well it means something wrong has happened in the meantime - since the moment we started the iteration. And hence this should be a legit error (it follows now the same logic as it is in the default `gfal-ls` or `gfal-rm` APIs). But in our case, since we are working all the time with an outdated list of files and directories coming from the RucioConsitencyMonitor (and in the worst case scenario we may end up with an age of up to 1 week), many of those directories may have gone by the time we  try to clean them, either because of a cleanup job has deleted them or because of a previous run of the unmerged service.. etc. So in this case I would think of considering this as  a 'previous' success (because the directory is already gone - which is our goal actually) rather than an error and would call it `True` in the next line. I would move this `error` message here just to `warning` and simply  continue. This way will avoid repeatedly retrying the deletions of all the files which has been there a week ago - meaning if we to not account for the fact that the directory is already gone (even though it may have not been us during the current cycle who removed it), then every time all the thousands of files which are still persisting in the consistency lists would trigger a series of deletions  doomed to failure every polling cycle.  

Please let me know what you think?

Reply:
Hi Todor, yes, I agree. If we cannot delete content in a given directory because the directory no longer exists, I think it's a good compromise to say that deletion of that directory is successful.

Reply:
Fixed!